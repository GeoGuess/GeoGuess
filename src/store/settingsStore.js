import * as MutationTypes from './mutation-types';
import { GAME_MODE, SCORE_MODE } from '../constants';
import i18n from '../lang';
import router from '../router';
import { getMaxDistanceBbox } from '../utils';
import bbox from '@turf/bbox';
import firebase from 'firebase/app';
import 'firebase/database';

export class GameSettings {
    constructor(
        _allPanorama = false,
        _timeLimitation = 0,
        _mode = GAME_MODE.CLASSIC,
        _timeAttack = false,
        _zoomControl = true,
        _moveControl = true,
        _panControl = true,
        _countdown = 0,
        _scoreMode = SCORE_MODE.NORMAL,
        _areaParams = null
    ) {
        this.allPanorama = _allPanorama;
        this.time = _timeLimitation;
        this.modeSelected = _mode;
        this.timeAttackSelected = _timeAttack;
        this.zoomControl = _zoomControl;
        this.moveControl = _moveControl;
        this.panControl = _panControl;
        this.countdown = _countdown;
        this.scoreMode = _scoreMode;
        this.areaParams = _areaParams;
    }
}

export default {
    namespaced: true,
    state: () => ({
        //Dialog
        isOpenDialogRoom: false,
        loadRoom: false,
        currentComponent: 'settingsMap',
        singlePlayer: true,

        // ROOM
        room: null,
        roomName: '',
        roomErrorMessage: null,
        playerNumber: 0,
        // SETTINGS
        gameSettings: new GameSettings(),
        players: [],
        name: localStorage.getItem('playerName') || '',
        invalidName: false,
    }),
    mutations: {
        [MutationTypes.SETTINGS_SET_ROOM](state, roomName) {
            state.room = firebase.database().ref(roomName);
            state.roomName = roomName;
            // Open Modal
            if (!state.isOpenDialogRoom) {
                state.loadRoom = true;
                state.isOpenDialogRoom = true;
            }

            state.room.once('value', (snapshot) => {
                if (snapshot.child('started').val()) {
                    state.roomErrorMessage = i18n.t(
                        'DialogRoom.alreadyStarted'
                    );
                    state.room.off();
                    return;
                }

                const numberOfPlayers = snapshot
                    .child('playerName')
                    .numChildren();
                const playerNumber = numberOfPlayers + 1;

                state.playerNumber = playerNumber;
                const name = state.name === '' ? i18n.t(
                                'CardRoomPlayerName.anonymousPlayerName'
                            ) + playerNumber : state.name;

                state.room.child('playerName/player'+playerNumber).onDisconnect().remove();

                
                if (numberOfPlayers === 0) {
                    // Put the tentative player's name into the room node
                    // So that other player can't enter as the first player while the player decide the name and room size
                    state.room.child('playerName').update(
                        {
                            player1: name,
                        },
                        (error) => {
                            if (!error) {
                                // Put the timestamp the room is created so the expired rooms can be removed by cloud function
                                state.room.update({
                                    createdAt:
                                        firebase.database.ServerValue.TIMESTAMP,
                                });
                                state.loadRoom = false;
                                state.currentComponent = 'settingsMap';
                            }
                        }
                    );
                } else {
                    // Put other player's tentative name
                    state.room
                        .child('playerName/player' + playerNumber)
                        .set(
                            name,
                            (error) => {
                                if (!error) {
                                    state.loadRoom = false;
                                    state.currentComponent = 'playerName';
                                }
                            }
                        );
                }
            });
        },
        [MutationTypes.SETTINGS_SET_ROOM_ERROR](state, error) {
            state.roomErrorMessage = error;
        },
        [MutationTypes.SETTINGS_SET_GAME_SETTINGS](state, settings) {
            if (settings.modeSelected) {
                settings.areaParams = null;
            }
            if (settings.areaParams) {
                settings.modeSelected = GAME_MODE.CUSTOM_AREA;
            }

            state.gameSettings = {
                ...state.gameSettings,
                ...settings,
            };
        },
        [MutationTypes.SETTINGS_SET_DIFFICULTY](state, difficulty){
            state.difficulty = difficulty;
        },
        [MutationTypes.SETTINGS_SET_BBOX](state, bbox){
            state.bboxObj = bbox;
        },
        [MutationTypes.SETTINGS_SET_OPEN_DIALOG_ROOM](state, open) {
            state.isOpenDialogRoom = open;
        },
        [MutationTypes.SETTINGS_SET_MODE_DIALOG_ROOM](state, singlePlayer) {
            state.singlePlayer = singlePlayer;
            state.currentComponent = singlePlayer ? 'settingsMap' : 'roomName';
        },

        [MutationTypes.SETTINGS_SET_STEP_DIALOG_ROOM](state, step) {
            state.currentComponent = step;
        },
        [MutationTypes.SETTINGS_SET_PLAYER_NAME](state, playerName) {
            state.invalidName = state.players.includes(playerName);
            if (!state.invalidName) {
                state.name = playerName;
                state.room
                    .child('playerName/player' + state.playerNumber)
                    .set(playerName);
            }
        },
        [MutationTypes.SETTINGS_SET_PLAYERS](state, players) {
            state.players = players;
        },
        [MutationTypes.SETTINGS_RESET](state) {
            state.room = null;
            state.roomName = '';
            state.playerNumber = 0;
            state.roomErrorMessage = null;
            state.players = [];
            state.gameSettings = new GameSettings();
        },
    },

    getters: {
        areasJson(state) {
            return state.areas;
        },
    },

    actions: {
        closeDialogRoom({ state, commit, dispatch }, cleanRoom = true) {
            commit(MutationTypes.SETTINGS_SET_OPEN_DIALOG_ROOM, false);

            // Remove the room
            if (state.room != null) {
                state.room.off();
                if (cleanRoom) {
                    if (state.playerNumber === 1) {
                        // Remove the entire node if the player is the first player
                        state.room.remove();
                    } else {
                        // Remove only the player's name node if the player isn't the first player
                        state.room
                            .child('playerName/player' + this.playerNumber)
                            .remove();
                    }
                }
            }

            dispatch('setMapLoaded', new Map(), { root: true });
            commit(MutationTypes.SETTINGS_RESET);
        },
        openDialogRoom({ commit }, isSinglePlayer = true) {
            commit(MutationTypes.SETTINGS_SET_MODE_DIALOG_ROOM, isSinglePlayer);
            commit(MutationTypes.SETTINGS_SET_OPEN_DIALOG_ROOM, true);
        },

        searchRoom({ commit, dispatch, state }, roomName) {
            commit(MutationTypes.SETTINGS_SET_MODE_DIALOG_ROOM, false);
            if (roomName == '') {
                commit(
                    MutationTypes.SETTINGS_SET_ROOM_ERROR,
                    i18n.t('DialogRoom.invalidRoomName')
                );
            } else {
                commit(MutationTypes.SETTINGS_SET_ROOM, roomName);
            }

            state.room.on('value', (snapshot) => {
                if (snapshot.child('playerName').exists())
                    state.players = Object.values(
                        snapshot.child('playerName').val()
                    );

                if (
                    state.currentComponent === 'playerName' &&
                    state.playerNumber !== 1 &&
                    snapshot.hasChild('size') &&
                    snapshot.hasChild('streetView')
                ) {
                    dispatch('startGame');
                }
            });
        },
        setSettings({ commit, state, rootState, dispatch }) {
            let difficulty = 2000;
            let bboxObj;
            if (rootState.homeStore.map.geojson) {
                bboxObj = bbox(rootState.homeStore.map.geojson);
                commit(MutationTypes.SETTINGS_SET_BBOX, bboxObj);

                difficulty = getMaxDistanceBbox(bboxObj) / 10;
            }
            commit(MutationTypes.SETTINGS_SET_DIFFICULTY, difficulty);
            if (!state.room) {
                router.push({
                    name: 'street-view',
                    params: {
                        ...state.gameSettings,
                        difficulty,
                        placeGeoJson: rootState.homeStore.map.geojson,
                        bboxObj: bboxObj,
                    },
                });
                dispatch('closeDialogRoom');
            } else {
                state.room.update(
                    {
                        ...state.gameSettings,
                        timeLimitation: state.gameSettings.time,
                        difficulty,
                        ...(rootState.homeStore.map.geojson && {placeGeoJson: rootState.homeStore.map.geojson}),
                        ...(bboxObj && { bboxObj: bboxObj }),
                    },
                    (error) => {
                        if (!error) {
                            commit(
                                MutationTypes.SETTINGS_SET_STEP_DIALOG_ROOM,
                                'playerName'
                            );
                        }
                    }
                );
            }
        },

        setPlayerName({ commit, state }, playerName) {
            if (playerName === '') {
                commit(
                    MutationTypes.SETTINGS_SET_PLAYER_NAME,
                    i18n.t('CardRoomPlayerName.anonymousPlayerName') +' '+ state.playerNumber
                );
            }else{
                localStorage.setItem('playerName', playerName);
                commit(MutationTypes.SETTINGS_SET_PLAYER_NAME, playerName);
            }

        },
        startGame({ state, dispatch, rootState }) {
            let gameParams = {};
            if (state.playerNumber === 1) {
                gameParams = {
                    ...state.gameSettings,
                    difficulty: state.difficulty,
                    placeGeoJson: rootState.homeStore.map.geojson,
                    bboxObj: state.bboxObj,
                };
                // Set flag started
                state.room.update({
                    size: state.players.length,
                    started: true,
                });
                dispatch('startGameMultiplayer', gameParams);
            } else {
                state.room.once('value', (snapshot) => {
                    gameParams = {
                        difficulty: snapshot.child('difficulty').val(),
                        bboxObj: snapshot.child('bboxObj').val(),
                        modeSelected: snapshot.child('modeSelected').val(),
                        timeAttackSelected: snapshot
                            .child('timeAttackSelected')
                            .val(),
                        zoomControl: snapshot.child('zoomControl').val(),
                        moveControl: snapshot.child('moveControl').val(),
                        panControl: snapshot.child('panControl').val(),
                        countdown: snapshot.child('countdown').val(),
                        allPanorama: snapshot.child('allPanorama').val(),
                        scoreMode: snapshot.child('scoreMode').val(),
                        areaParams: snapshot.child('areaParams').val(),
                    };

                    dispatch('startGameMultiplayer', gameParams);
                });
            }
        },
        startGameMultiplayer({ state, rootState, dispatch }, gameParams) {
            // Start the game
            router.push({
                name: 'with-friends',
                params: {
                    ...gameParams,
                    roomName: state.roomName,
                    playerName: state.name,
                    playerNumber: state.playerNumber,
                    placeGeoJson: rootState.homeStore.map.geojson,
                    multiplayer: true,
                },
            });

            dispatch('closeDialogRoom', false);
        },
    },
};
