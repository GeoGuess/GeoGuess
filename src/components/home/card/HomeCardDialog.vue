<template>
    <v-dialog
        v-model="visible"
        max-width="500"
        :fullscreen="$viewport.width < 450"
    >
        <template v-slot:activator="{ on, attrs }">
            <v-btn v-bind="attrs" text color="darkGreen" v-on="on">
                {{ $t('Home.play') }}
            </v-btn>
        </template>
        <v-card>
            <v-img
                class="white--text align-end"
                height="230px"
                gradient="rgba(0,0,0,0), rgba(0,0,0,0.8)"
                :src="
                    data.imageUrl ||
                        `https://source.unsplash.com/500x230/weekly?${encodeURI(
                            data.nameLocate
                        )}`
                "
            >
                <v-card-title>{{ data.nameLocate }}</v-card-title>
            </v-img>
            <v-card-subtitle class="pt-3 pb-2 font-italic">
                {{ data.author }}
            </v-card-subtitle>
            <v-card-text class="text--primary map-dialog__description">
                {{ data.descriptionLocate }}
            </v-card-text>
            <v-card-actions>
                <v-btn color="error" text @click="visible = false">
                    {{ $t('cancel') }}
                </v-btn>

                <v-spacer />
                <v-btn color="primary" @click="onClickSinglePlayer">
                    {{ $t('DialogRoom.singlePlayer') }}
                </v-btn>

                <v-btn color="secondary" dark @click="onClickMultiPlayer">
                    {{ $t('DialogRoom.withFriends') }}
                </v-btn>
            </v-card-actions>
        </v-card>
    </v-dialog>
</template>

<script>
import { mapActions, mapMutations } from 'vuex';
import { SETTINGS_SET_GAME_SETTINGS } from '../../../store/mutation-types';
export default {
    props: {
        data: Object,
        type: {
            type: String,
            validator: (v) => ['map', 'area'].includes(v),
        },
    },
    data() {
        return {
            visible: false,
        };
    },
    methods: {
        ...mapMutations('settingsStore', {
            setGameSettings: SETTINGS_SET_GAME_SETTINGS,
        }),
        ...mapActions('settingsStore', ['openDialogRoom']),
        ...mapActions(['loadGeoJsonFromUrl', 'setGeoJson', 'setMapLoaded']),
        setMap() {
            if (this.type === 'area') {
                this.loadGeoJsonFromUrl(this.data.data.urlArea);
                this.setGameSettings({ areaParams: this.data });
            } else {
                if (this.data.custom) {
                    this.setMapLoaded(this.data);
                } else {
                    this.loadGeoJsonFromUrl(this.data.url);
                }
            }
            this.visible = false;
        },
        onClickSinglePlayer() {
            this.setMap();
            this.openDialogRoom(true);
        },
        onClickMultiPlayer() {
            this.setMap();
            this.openDialogRoom(false);
        },
    },
};
</script>

<style scoped>
.map-dialog__description {
    max-height: 20vh;
    overflow-y: auto;
}
</style>
