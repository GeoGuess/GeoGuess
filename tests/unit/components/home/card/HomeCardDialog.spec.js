import HomeCardDialog from '@/components/home/card/HomeCardDialog.vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import settingsStore from '@/store/modules/settings.store';
import appInit from '../../../testutils/appInit';

const args = appInit(createLocalVue());

describe('HomeCardDialog.vue', () => {
    let store, actions, loadMap;
    beforeEach(() => {
        actions = {
            openDialogRoom: jest.fn(),
        };
        loadMap = jest.fn();
        store = new Vuex.Store({
            modules: {
                settingsStore: {
                    namespaced: true,
                    state: settingsStore.state,
                    getters: settingsStore.getters,
                    mutations: settingsStore.mutations,
                    actions,
                },
                homeStore:{
                    actions:{
                        loadMap                        
                    }
                }
            },
        });
    });
    it('methods', () => {
        const map = {
            name: {
                en: 'Biggest City',
                fr: 'Grande Villes',
            },
            nameLocate: 'Grande Villes',
            description: {
                fr:
                    'Une liste de plus de 40 villes : PARIS, KOBE OSAKA, SEOUL, HOUSTON, BARCELONA, PHILADELPHIA, SANTIAGO, LAGOS, DALLAS, NEW YORK, ISTANBUL, TOKYO, FUKUOKA, LONDON, KUALA LUMPUR, LIMA, HO CHI MINH CITY, MANILA, GUADALAJARA, MADRID, NAGOYA, SINGAPORE, JOHANNESBURG, BELO HORIZONTE, TORONTO, MEXICO CITY, MIAMI, ATLANTA, RIO DE JANEIRO, BUENOS AIRES, SAO PAULO, CHICAGO,…',
                en:
                    'List of 40 biggest cities of the world: PARIS, KOBE OSAKA, SEOUL, HOUSTON, BARCELONA, PHILADELPHIA, SANTIAGO, LAGOS, DALLAS, NEW YORK, ISTANBUL, TOKYO, FUKUOKA, LONDON, KUALA LUMPUR, LIMA, HO CHI MINH CITY, MANILA, GUADALAJARA, MADRID, NAGOYA, SINGAPORE, JOHANNESBURG, BELO HORIZONTE, TORONTO, MEXICO CITY, MIAMI, ATLANTA, RIO DE JANEIRO, BUENOS AIRES, SAO PAULO, CHICAGO,…',
            },
            descriptionLocate:
                'Une liste de plus de 40 villes : PARIS, KOBE OSAKA, SEOUL, HOUSTON, BARCELONA, PHILADELPHIA, SANTIAGO, LAGOS, DALLAS, NEW YORK, ISTANBUL, TOKYO, FUKUOKA, LONDON, KUALA LUMPUR, LIMA, HO CHI MINH CITY, MANILA, GUADALAJARA, MADRID, NAGOYA, SINGAPORE, JOHANNESBURG, BELO HORIZONTE, TORONTO, MEXICO CITY, MIAMI, ATLANTA, RIO DE JANEIRO, BUENOS AIRES, SAO PAULO, CHICAGO,…',
            author: 'BilelJegham',
            imageSrc:
                'https://images.unsplash.com/photo-1477959858617-67f85cf4f1df?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=140',
            url: 'https://mapurl.geojson',
        };
        const wrapper = shallowMount(HomeCardDialog, {
            ...args,
            store,
            propsData: {
                data: map,
            },
        });
        wrapper.setData({ visible: true });
        wrapper.vm.onClickSinglePlayer();

        expect(wrapper.vm.visible).toEqual(false);
        expect(actions.openDialogRoom).toBeCalled();
        expect(loadMap).toBeCalledWith(expect.anything(), map);

        wrapper.setData({ visible: true });
        wrapper.vm.onClickMultiPlayer();

        expect(wrapper.vm.visible).toEqual(false);

        expect(actions.openDialogRoom).toBeCalled();

        expect(loadMap).toBeCalledTimes(2);

    });
});
