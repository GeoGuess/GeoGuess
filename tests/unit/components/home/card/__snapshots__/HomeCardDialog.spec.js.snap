// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeCardDialog.vue methods 1`] = `
<v-dialog-stub opendelay="0" closedelay="0" openonclick="true" contentclass="" maxwidth="500" origin="center center" retainfocus="true" transition="dialog-transition">
  <v-card-stub loaderheight="4" tag="div">
    <v-img-stub height="230px" gradient="rgba(0,0,0,0), rgba(0,0,0,0.8)" options="[object Object]" position="center center" src="https://images.unsplash.com/photo-1477959858617-67f85cf4f1df?ixlib=rb-1.2.1&amp;ixid=eyJhcHBfaWQiOjEyMDd9&amp;auto=format&amp;fit=crop&amp;w=300&amp;q=140" transition="fade-transition" class="white--text align-end">
      <v-card-title-stub tag="div">Grande Villes</v-card-title-stub>
    </v-img-stub>
    <v-card-subtitle-stub tag="div" class="pt-3 pb-2 font-italic">
      BilelJegham
    </v-card-subtitle-stub>
    <v-card-text-stub tag="div" class="text--primary map-dialog__description">
      Une liste de plus de 40 villes : PARIS, KOBE OSAKA, SEOUL, HOUSTON, BARCELONA, PHILADELPHIA, SANTIAGO, LAGOS, DALLAS, NEW YORK, ISTANBUL, TOKYO, FUKUOKA, LONDON, KUALA LUMPUR, LIMA, HO CHI MINH CITY, MANILA, GUADALAJARA, MADRID, NAGOYA, SINGAPORE, JOHANNESBURG, BELO HORIZONTE, TORONTO, MEXICO CITY, MIAMI, ATLANTA, RIO DE JANEIRO, BUENOS AIRES, SAO PAULO, CHICAGO,â€¦
    </v-card-text-stub>
    <v-card-actions-stub tag="div">
      <v-btn-stub color="error" tag="button" activeclass="" text="true" type="button">
        CANCEL
      </v-btn-stub>
      <v-spacer-stub tag="div"></v-spacer-stub>
      <v-btn-stub color="primary" tag="button" activeclass="" type="button">
        Single Player
      </v-btn-stub>
      <v-btn-stub color="secondary" dark="true" tag="button" activeclass="" type="button">
        With Friends
      </v-btn-stub>
    </v-card-actions-stub>
  </v-card-stub>
</v-dialog-stub>
`;
